docs/config_file.rst
.. _config file:

Config file
===========

Users from different instruments should use the corresponding config file to config
the instrument specific data sources, the detector specific setups and so on.

Each instrument has a default config file, which can be found in the
`github repository <https://github.com/European-XFEL/EXtra-foam/tree/dev/extra_foam/configs>`__.
We appreciate if the beamline scientists can help us keep the default config file updated.

Let's take *FXE* for example, when one starts a detector on topic *FXE* for the first by
time:

.. code-block:: bash

    extra-foam LPD FXE

, the system will create a new config file `$HOME/.EXtra-foam/fxe.config.yaml` using the default one.
The first block of the config file looks like the following:

.. code-block:: yaml

    SOURCE:
        # Default source type: FILES or BRIDGE
        DEFAULT_TYPE: 1
        CATEGORY:
            LPD:
                PIPELINE:
                    FXE_DET_LPD1M-1/CAL/APPEND_CORRECTED:
                        - image.data
                    FXE_DET_LPD1M-1/DET/*CH0:xtdf:
                        - image.data

            JungFrau:
                PIPELINE:
                    FXE_XAD_JF1M/DET/RECEIVER-1:display:
                        - data.adc
                    FXE_XAD_JF1M/DET/RECEIVER-2:display:
                        - data.adc

            XGM:
                CONTROL:
                    SA1_XTD2_XGM/DOOCS/MAIN:
                        - pulseEnergy.photonFlux
                        - beamPosition.ixPos
                        - beamPosition.iyPos
                PIPELINE:
                    SA1_XTD2_XGM/DOOCS/MAIN:output:
                        - data.intensitySa1TD

            Motor:
                CONTROL:
                    FXE_SMS_USR/MOTOR/UM01:
                        - actualPosition
                    FXE_SMS_USR/MOTOR/UM02:
                        - actualPosition


The next block of the config file looks like the following:

.. code-block:: yaml

    DETECTOR:
        LPD:
            GEOMETRY_FILE: lpd_mar_18_axesfixed.h5
            # - For lpd_mar_18.h5 and LPDGeometry in karabo_data
            # "QUAD_POSITIONS": [[-13.0, -299.0], [11.0, -8.0], [-254.0, 16.0], [-278.0, -275.0]],
            # - For lpd_mar18_axesfixed.h5 and LPD_1MGeometry in karabo_data
            # The geometry uses XFEL standard coordinate directions.
            QUAD_POSITIONS:
                x1:   11.4
                y1:  299
                x2:  -11.5
                y2:    8
                x3:  254.5
                y3:  -16
                x4:  278.5
                y4:  275
            BRIDGE_ADDR: 10.253.0.53
            BRIDGE_PORT: 4501
            LOCAL_ADDR: 127.0.0.1
            LOCAL_PORT: 45451
            SAMPLE_DISTANCE": 0.4
            PHOTON_ENERGY": 9.3

        JungFrau:
            BRIDGE_ADDR: 10.253.0.53
            BRIDGE_PORT: 4501
            LOCAL_ADDR: 127.0.0.1
            LOCAL_PORT: 45453
            SAMPLE_DISTANCE: 2.0
            PHOTON_ENERGY: 9.3


docs/configuration.rst
.. code-block:: yaml

    DSSC: # detector name
        # <Karabo device ID> + space + <property>
        FXE_DET_LPD1M-1/CAL/APPEND_CORRECTED image.data
        FXE_DET_LPD1M-1/DET/*CH0:xtdf image.data
        # absolute path of geometry file
        GEOMETRY_FILE: geometries/lpd_mar_18_axesfixed.h5
        # quadrant coordinates for assembling detector modules
        QUAD_POSITIONS:
          x1:   11.4
          y1:  299
          x2:  -11.5
          y2:    8
          x3:  254.5
          y3:  -16
          x4:  278.5
          y4:  275
        # TCP address of the ZMQ bridge
        BRIDGE_ADDR": 10.253.0.53
        # TCP port of the ZMQ bridge
        BRIDGE_PORT": 4501
        # TCP address of the file streamer
        LOCAL_ADDR: 127.0.0.1
        # TCP port of the file streamer
        LOCAL_PORT: 45454
        # distance from sample to detector plan (orthogonal distance,
        # not along the beam), in meter
        SAMPLE_DISTANCE": 0.4
        # photon energy, in keV
        PHOTON_ENERGY": 9.3

.. code-block:: yaml

    XGM:
        # For control data: <Karabo device ID> <property>
        SA1_XTD2_XGM/DOOCS/MAIN pulseEnergy.photonFlux
        SA1_XTD2_XGM/DOOCS/MAIN beamPosition.ixPos
        SA1_XTD2_XGM/DOOCS/MAIN beamPosition.iyPos
        # For pipeline data: <Karabo device ID>:<output channel>
        SA1_XTD2_XGM/DOOCS/MAIN:output:
            property:
                - data.intensityTD
                - data.intensitySa1TD
                - data.intensitySa3TD
            vrange:
                - 0.0
                - .inf

.. code-block:: yaml

    Motor: # must be Motor
        # <Karabo device ID> + space + <property>
        FXE_SMS_USR/MOTOR/UM01 actualPosition
        FXE_SMS_USR/MOTOR/UM02 actualPosition
        FXE_SMS_USR/MOTOR/UM04 actualPosition:
            vrange:
                - 0.0
                - .inf

docs/changelog.rst
Changelog
=========

1.0.0 (31 July 2020)
------------------------

- **Improvement**

    - Improve FFT of image in feature extraction. :ghpull:`268`
    - Improve fitting interface. :ghpull:`265`
    - Add reset panel in the main GUI. :ghpull:`264`
    - Update default config files. :ghpull:`262`
    - Improve azimuthal integration performance. :ghpull:`261`, :ghpull:`268`
    - Revert y-axis orientation change. :ghpull:`260`
    - Add Python binding for double dtype in generalized geometry. :ghpull:`259`

- **New Feature**

    - Implemented pulse-resolved correlation in special suite. :ghpull:`270`
    - Add curve fitting in histogram window. :ghpull:`267`
    - Support streaming Basler camera from files. :ghpull:`266`
    - Add more fitting types, Gaussian, Lorentzian and erf. :ghpull:`265`
    - Add pump-probe FOM to histogram window. :ghpull:`263`
    - Reset moving average in correlation scan mode. :ghpull:`263`


0.9.1 (15 July 2020)
------------------------

- **Bug Fix**

    - Fix transform type update in TransformView in ImageTool. :ghpull:`251`

- **Improvement**

    - PlotItem will not be shown in legend if its name is empty. :ghpull:`254`
    - Improve y2-axis plot implementation. :ghpull:`253`
    - Improve stack detector modules. :ghpull:`247`
    - Implement ScatterPlotItem to replace the pyqtgraph one. :ghpull:`238`

- **New Feature**

    - Implement curve fitting in correlation window. :ghpull:`255`
    - Implement azimuthal integration and concentric ring detection in C++. :ghpull:`252`


0.9.0 (30 June 2020)
------------------------

- **Bug Fix**

    - Fix bug in data source tree introduced in 0.8.4. :ghpull:`235`
    - Fix transparent Nan pixel with the latest pyqtgraph version. :ghpull:`231`

- **Improvement**

    - Improve performance of PlotCurveItem and add benchmarks for PlotItems and ImageViewF. :ghpull:`243`
    - Make ctrl widgets in Imagetool more compact. :ghpull:`241`
    - Improve C++ code quality and suppress TBB deprecate warning. :ghpull:`239`
    - Add meter bar to plot area. :ghpull:`236`
    - Reimplement ImageItem and PlotItem to replace the pyqtgraph ones. :ghpull:`232`, :ghpull:`242`
    - Improve data source updating and color encoding matched source items in
      the data source tree. :ghpull:`230`, :ghpull:`234`
    - Rename configurator to analysis setup manager and allow take snapshot for
      each setup. :ghpull:`228`
    - Update data source and trouble shooting in documentation. :ghpull:`227`
    - Add summary of compiler flags for EXtra-foam-python. :ghpull:`226`
    - Update installation (gcc7) and CI (gcc8). :ghpull:`226`
    - Rename misleading mouse mode in the right-click context menu. :ghpull:`211`
    - Update pyqtgraph to the latest master branch. :ghpull:`206`, :ghpull:`211`, :ghpull:`233`

- **New Feature**

    - Annotate peaks in azimuthal integration view. :ghpull:`240`
    - Enable Legend for all the plot items. :ghpull:`206`, :ghpull:`237`
    - Implement logarithmic X/Y scale in right-click context menu. :ghpull:`206`
    - Enable C++ API installation and add examples. :ghpull:`227`


0.8.4 (8 June 2020)
------------------------

- **Bug Fix**

    - Beam size in the bar plot will change when the resolution changes in the
      correlation analysis. :ghpull:`209`

- **Improvement**

    - Update Redis versions: server -> 6.0.5; redis-py -> 3.5.2; hiredis-py -> 1.0.1. :ghpull:`220`
    - Slightly improve image proc C++ code performance on machines with few threads. :ghpull:`219`
    - Visualize data type in data source Tree. :ghpull:`218`
    - Improve the performance of pulse filter. :ghpull:`214`
    - Improve setup.py. :ghpull:`212`
    - Keep data when resolution changes in correlation analysis; move sequence classes
      into algorithms/data_structures. :ghpull:`209`
    - Mask ASIC edge when n_modules == 1; simplify geometry binding code. :ghpull:`207`
    - Add benchmark for generalized geometry. :ghpull:`205`
    - Make special suite more self-contained. :ghpull:`204`
    - Mask JungFrau and JungFrauPR. :ghpull:`200`
    - Move statistics ctrl widgets into corresponding windows. :ghpull:`199`

- **New Feature**

    - Color encoding matched sources in the data source tree. :ghpull:`220`
    - Add processed pulse counter in ImageTool. :ghpull:`216`
    - Add support for C++ API. :ghpull:`213`
    - Add xtensor-blas as a submodule. :ghpull:`210`
    - Implement image transform processor and view (FFT, edge detection). :ghpull:`203`
    - Integrate Karabo gate (PipeToEXtraFoam device) which allows request pipeline

extra_foam/config.py
"""
Distributed under the terms of the BSD 3-Clause License.

The full license is in the file LICENSE, distributed with this software.

Author: Jun Zhu <jun.zhu@xfel.eu>
Copyright (C) European X-Ray Free-Electron Laser Facility GmbH.
All rights reserved.
"""
from enum import IntEnum
import os.path as osp
import shutil
from collections import abc, namedtuple, OrderedDict

import yaml
from yaml.scanner import ScannerError
from yaml.parser import ParserError

import numpy as np

from . import ROOT_PATH
from .logger import logger


_MAX_INT32 = np.iinfo(np.int32).max
_MIN_INT32 = np.iinfo(np.int32).min


class DataSource(IntEnum):
    FILE = 0  # data from files (run directory)
    BRIDGE = 1  # real-time data from the bridge
    UNKNOWN = 2  # not specified


class KaraboType(IntEnum):
    CONTROL_DATA = 0
    PIPELINE_DATA = 1


class PumpProbeMode(IntEnum):
    UNDEFINED = 0
    REFERENCE_AS_OFF = 1  # use pre-defined reference image
    SAME_TRAIN = 2  # on-/off- pulses in the same train
    EVEN_TRAIN_ON = 3  # on-/off- pulses have even/odd train IDs, respectively
    ODD_TRAIN_ON = 4  # on/-off- pulses have odd/even train IDs, respectively


class RoiCombo(IntEnum):
    UNDEFINED = 0
    ROI1 = 1
    ROI2 = 2
    ROI1_SUB_ROI2 = 3
    ROI1_ADD_ROI2 = 4
    ROI1_DIV_ROI2 = 5
    ROI3 = 11
    ROI4 = 12
    ROI3_SUB_ROI4 = 13
    ROI3_ADD_ROI4 = 14


class RoiFom(IntEnum):
    SUM = 1
    MEAN = 2
    MEDIAN = 3
    MAX = 4
    MIN = 5
    STD = 6  # standard deviation
    VAR = 7  # variance
    N_STD = 16  # normalized standard deviation (speckle contrast)
    N_VAR = 17  # normalized variance (speckle contrast)


class RoiProjType(IntEnum):
    SUM = 1
    MEAN = 2


class AnalysisType(IntEnum):
    UNDEFINED = 0
    PUMP_PROBE = 1
    ROI_FOM = 11
    ROI_NORM = 12
    ROI_PROJ = 21
    AZIMUTHAL_INTEG = 41
    AZIMUTHAL_INTEG_PEAK = 42
    AZIMUTHAL_INTEG_PEAK_Q = 43
    AZIMUTHAL_INTEG_COM = 44
    PULSE = 2700
    PUMP_PROBE_PULSE = 2701
    ROI_FOM_PULSE = 2711
    ROI_NORM_PULSE = 2712
    ROI_PROJ_PULSE = 2721
    AZIMUTHAL_INTEG_PULSE = 2741


class GeomAssembler(IntEnum):
    OWN = 1  # use Extra-foam own geometry assembler
    EXTRA_GEOM = 2  # use Extra-geom geometry assembler


class ImageTransformType(IntEnum):
    CONCENTRIC_RINGS = 0
    FOURIER_TRANSFORM = 1
    EDGE_DETECTION = 2
    UNDEFINED = 255


class PipelineSlowPolicy(IntEnum):
    DROP = 0
    WAIT = 1


class CalibrationOffsetPolicy(IntEnum):
    UNDEFINED = 0
    INTRA_DARK = 1


_user_analysis_types = OrderedDict({
    "": AnalysisType.UNDEFINED,
    "pump-probe": AnalysisType.PUMP_PROBE,
    "ROI FOM": AnalysisType.ROI_FOM,
    "ROI proj": AnalysisType.ROI_PROJ,
    "azimuthal integ (sum)": AnalysisType.AZIMUTHAL_INTEG,
    "azimuthal integ (peak)": AnalysisType.AZIMUTHAL_INTEG_PEAK,
    "azimuthal integ (peak q)": AnalysisType.AZIMUTHAL_INTEG_PEAK_Q,
    "azimuthal integ (CoM)": AnalysisType.AZIMUTHAL_INTEG_COM
})


def get_analysis_types(without=None):
    if without is not None:
        return { k: v for k, v in _user_analysis_types.items()
                 if v != without }
    else:
        return _user_analysis_types


def list_azimuthal_integ_methods(detector):
    """Return a list of available azimuthal integration methos.

    :param str detector: detector name
    """
    if detector in ['AGIPD', 'DSSC', 'LPD']:
        return ['BBox', 'splitpixel', 'csr', 'nosplit_csr', 'csr_ocl',
                'lut', 'lut_ocl']
    return ['nosplit_csr', 'csr_ocl', 'csr', 'BBox', 'splitpixel',
            'lut', 'lut_ocl']


_PlotLabelItem = namedtuple("_PlotLabel", ['x', 'y'])


class PlotLabel(abc.Mapping):
    """Labels used in data visualization."""
    _labels = {
        AnalysisType.ROI_PROJ: _PlotLabelItem("x", "Projection"),
        AnalysisType.AZIMUTHAL_INTEG: _PlotLabelItem(
            "Momentum transfer (1/A)", "Scattering signal (arb. u.)")
    }

    def __init__(self):
        super().__init__()

        for item in AnalysisType:
            if item not in self._labels:
                self._labels[item] = _PlotLabelItem("", "")

    def __contains__(self, analysis_type):
        """Override."""
        return analysis_type in self._labels

    def __getitem__(self, analysis_type):
        """Override."""
        return self._labels[analysis_type]

    def __iter__(self):
        """Override."""
        return iter(self._labels)

    def __len__(self):
        """Override."""
        return len(self._labels)


plot_labels = PlotLabel()


class BinMode(IntEnum):
    ACCUMULATE = 0
    AVERAGE = 1


class Normalizer(IntEnum):
    UNDEFINED = 0
    AUC = 1  # area under curve
    XGM = 2
    DIGITIZER = 3
    ROI = 4


# a simple class saves the trouble when the attribute needs to be read/write
# from/at Redis.
class MaskState:
    UNMASK = 0
    MASK = 1
    CLEAR_MASK = -1


class _Config(dict):
    """Config implementation."""

    _abs_dirpath = osp.abspath(osp.dirname(__file__))

    _opts = {
        # instrument name
        "TOPIC": "",
        # -------------------------------------------------------------
        # Main detector
        # -------------------------------------------------------------
        # main detector name
        "DETECTOR": "",
        # Whether pulse-resolved analysis setup is used
        "PULSE_RESOLVED": False,
        # whether a geometry is required to assemble the detector modules
        "REQUIRE_GEOMETRY": False,
        # absolute path of geometry file
        "GEOMETRY_FILE": "",
        # Note: due to the historical reason, we have "QUAD_POSITIONS" in the
        #       config file for a long time. However, it is only used in
        #       1M detectors geometry, while "MODULE_POSITIONS", which is
        #       introduced in version 0.8.3, is used in general detector
        #       geometry.
        # quadrant coordinates for assembling detector modules,
        # ((x1, y1), (x2, y2), (x3, y3), (x4, y4))
        "QUAD_POSITIONS": ((0, 0), ) * 4,
        # module coordinates for assembling detector modules,
        # a tuple of (x, y), we assume 12 modules will be used at maximum
        "MODULE_POSITIONS": ((0, 0),) * 12,
        # number of modules
        "NUMBER_OF_MODULES": 1,
        # shape of a single module
        "MODULE_SHAPE": (-1, -1),
        # detector pixel size, in meter
        "PIXEL_SIZE": 1.e-3,
        # whether the detector supports masking edge pixels of tiles
        "MASK_TILE_EDGE": False,
        # whether the detector supports masking edge pixels of asics
        "MASK_ASIC_EDGE": False,
        # Default TCP address when data source is BRIDGE
        "BRIDGE_ADDR": "127.0.0.1",
        # Default TCP port when data source is BRIDGE
        "BRIDGE_PORT": 4501,
        # Default TCP address when data source is FILE
        "LOCAL_ADDR": "127.0.0.1",
        # Default TCP port when data source is FILE
        "LOCAL_PORT": 45454,
        # distance from sample to detector plan (orthogonal distance,
        # not along the beam), in meter
        "SAMPLE_DISTANCE": 1.0,
        # photon energy, in keV
        "PHOTON_ENERGY": 12.4,
        # -------------------------------------------------------------
        # Process
        # -------------------------------------------------------------
        # timeout when cleaning up remnant processes, in second
        "PROCESS_CLEANUP_TIMEOUT": 1,
        # interval for updating the status of all processes, in milliseconds
        "PROCESS_MONITOR_UPDATE_TIMER": 5000,
        # -------------------------------------------------------------
        # Pipeline
        # -------------------------------------------------------------
        # maximum length of the queue in data pipeline (the smaller the queue
        # size, the smaller the latency)
        "PIPELINE_MAX_QUEUE_SIZE": 2,
        "PIPELINE_SLOW_POLICY": PipelineSlowPolicy.DROP,
        # timeout of the zmq bridge, in second
        "BRIDGE_TIMEOUT": 0.1,
        # default extension port
        "EXTENSION_PORT": 5555,
        # size of cache when correlating data arriving at different time
        # with train ID
        "TRANSFORMER_CACHE_SIZE": 20,
        # -------------------------------------------------------------
        # Source of data
        # -------------------------------------------------------------
        # default source type: online or from file
        "SOURCE_DEFAULT_TYPE": int(DataSource.BRIDGE),
        # dtype of the processed image data
        # Note: the cpp code must have the corresponding overload!!!
        #       np.float32 -> float
        #       np.float64 -> double
        "SOURCE_PROC_IMAGE_DTYPE": np.float32,
        # dtype of the raw image data
        "SOURCE_RAW_IMAGE_DTYPE": np.uint16,
        # interval for updating available data sources, in milliseconds
        "SOURCE_AVAIL_UPDATE_TIMER": 1000,
        # After how long the available sources key expires, in seconds
        "SOURCE_EXPIRATION_TIMER": 2,
        "SOURCE_USER_DEFINED_CATEGORY": "User-defined",
        # -------------------------------------------------------------
        # REDIS
        # -------------------------------------------------------------
        # absolute path of the Redis server executable
        "REDIS_EXECUTABLE": osp.join(_abs_dirpath, "thirdparty/bin/redis-server"),
        # default Redis port used in testing. Each detector has its
        # dedicated Redis port.
        "REDIS_PORT": 6379,
        # maximum attempts to ping the Redis server before shutting down the app
        "REDIS_MAX_PING_ATTEMPTS": 3,
        # interval for pinging the Redis server from the main GUI, in milliseconds
        "REDIS_PING_ATTEMPT_INTERVAL": 5000,
        # If the path is given, redis-py will use UnixDomainSocketConnection.
        # Otherwise, normal TCP socket connection.
        "REDIS_UNIX_DOMAIN_SOCKET_PATH": "",
        # maximum allowed REDIS memory (fraction of system memory)
        "REDIS_MAX_MEMORY_FRAC": 0.2,  # must <= 0.5
        # password to access the Redis server
        "REDIS_PASSWORD": "karaboFAI",  # FIXME
        "REDIS_LOGFILE": osp.join(ROOT_PATH, "redis.log"),
        # -------------------------------------------------------------
        # GUI
        # -------------------------------------------------------------
        # interval for polling new processed data, in milliseconds
        "GUI_PLOT_UPDATE_TIMER": 10,
        # interval for updating plots with state, in milliseconds
        # (for example, correlation, heatmap, etc.)
        "GUI_PLOT_WITH_STATE_UPDATE_TIMER": 1000,
        # initial (width, height) of the main GUI window
        "GUI_MAIN_GUI_SIZE": (1720, 1080),
        # initial (width, height) of the image tool window
        "GUI_IMAGE_TOOL_SIZE": (1720, 1080),
        # initial (width, height) of a large plot window
        "GUI_PLOT_WINDOW_SIZE": (1520, 1080),
        # color map in contour plots, valid options are: thermal, flame,
        # yellowy, bipolar, spectrum, cyclic, greyclip, grey, viridis,
        # inferno, plasma, magma
        "GUI_COLOR_MAP": 'plasma',
        # foreground/background color (r, g, b, alpha)
        "GUI_FOREGROUND_COLOR": (0, 0, 0, 255),
        "GUI_BACKGROUND_COLOR": (225, 225, 225, 255),
        # colors of for ROI bounding boxes 1 to 4
        "GUI_ROI_COLORS": ('b', 'r', 'g', 'o'),
        # colors (master, slave) for correlation plots 1 and 2
        "GUI_CORRELATION_COLORS": (('b', 'r'), ('g', 'p')),
        # color of the image mask bounding box while drawing
        "GUI_MASK_BOUNDING_BOX_COLOR": 'b',
        # color of the masked area for MaskItem
        "GUI_MASK_FILL_COLOR": 'g',
        # -------------------------------------------------------------
        # Misc
        # -------------------------------------------------------------
        # max number of pulses per pulse train
        "MAX_N_PULSES_PER_TRAIN": 2700,
    }

    _AreaDetectorConfig = namedtuple("_AreaDetectorConfig", [
        "REDIS_PORT", "PULSE_RESOLVED", "REQUIRE_GEOMETRY",
        "NUMBER_OF_MODULES", "MODULE_SHAPE", "PIXEL_SIZE",
        "MASK_TILE_EDGE", "MASK_ASIC_EDGE",
    ])

    # "name" is only used for labeling, it is ignored in the pipeline code.
    StreamerEndpointItem = namedtuple("StreamerEndpointItem",
                                      ["name", "type", "address", "port"])

    for key in _AreaDetectorConfig._fields:
        assert key in _opts

    _detector_config = {
        "AGIPD": _AreaDetectorConfig(
            REDIS_PORT=6378,
            PULSE_RESOLVED=True,
            REQUIRE_GEOMETRY=True,
            NUMBER_OF_MODULES=16,
            MODULE_SHAPE=(512, 128),
            PIXEL_SIZE=0.2e-3,
            MASK_TILE_EDGE=True,
            MASK_ASIC_EDGE=False,
        ),
        "LPD": _AreaDetectorConfig(
            REDIS_PORT=6379,
            PULSE_RESOLVED=True,
            REQUIRE_GEOMETRY=True,
            NUMBER_OF_MODULES=16,
            MODULE_SHAPE=(256, 256),
            PIXEL_SIZE=0.5e-3,
            MASK_TILE_EDGE=True,
            MASK_ASIC_EDGE=False,
        ),
        # DSSC has hexagonal pixels:
        # 236 μm step in fast-scan axis, 204 μm in slow-scan
        "DSSC": _AreaDetectorConfig(
            REDIS_PORT=6380,
            PULSE_RESOLVED=True,
            REQUIRE_GEOMETRY=True,
            NUMBER_OF_MODULES=16,
            MODULE_SHAPE=(128, 512),
            PIXEL_SIZE=0.22e-3,
            MASK_TILE_EDGE=True,
            MASK_ASIC_EDGE=False,
        ),
        "JungFrau": _AreaDetectorConfig(
            REDIS_PORT=6381,
            PULSE_RESOLVED=True,
            REQUIRE_GEOMETRY=False,
            NUMBER_OF_MODULES=1,
            MODULE_SHAPE=(512, 1024),
            PIXEL_SIZE=0.075e-3,
            MASK_TILE_EDGE=False,
            MASK_ASIC_EDGE=True,
        ),
        "FastCCD": _AreaDetectorConfig(
            REDIS_PORT=6383,
            PULSE_RESOLVED=False,
            REQUIRE_GEOMETRY=False,
            NUMBER_OF_MODULES=1,
            MODULE_SHAPE=(1934, 960),
            PIXEL_SIZE=0.030e-3,
            MASK_TILE_EDGE=False,
            MASK_ASIC_EDGE=False,
        ),
        "ePix100": _AreaDetectorConfig(
            REDIS_PORT=6384,
            PULSE_RESOLVED=False,
            REQUIRE_GEOMETRY=False,
            NUMBER_OF_MODULES=1,
            MODULE_SHAPE=(708, 768),
            PIXEL_SIZE=0.050e-3,
            MASK_TILE_EDGE=False,
            MASK_ASIC_EDGE=True,
        ),
        "BaslerCamera": _AreaDetectorConfig(
            REDIS_PORT=6389,
            PULSE_RESOLVED=False,
            REQUIRE_GEOMETRY=False,
            NUMBER_OF_MODULES=1,
            # BaslerCamera has quite a few different models with different
            # module shapes and pixel sizes.
            MODULE_SHAPE=(-1, -1),
            PIXEL_SIZE=0.0022e-3,
            MASK_TILE_EDGE=False,
            MASK_ASIC_EDGE=False,
        ),
    }

    _misc_source_categories = (
        'XGM', 'Digitizer', 'Magnet', 'Motor', 'Monochromator')

    def __init__(self):
        super().__init__()

        self.update(self._opts)

        # ctrl data sources listed in the DataSourceWidget
        self.control_sources = dict()
        # pipeline data sources listed in the DataSourceWidget
        self.pipeline_sources = dict()

        self.appendix_streamers = []

    @classmethod
    def topics(cls):
        return 'SPB', 'FXE', 'SCS', 'SQS', 'MID', 'HED'

    @classmethod
    def detectors(cls):
        return tuple(cls._detector_config.keys())

    @classmethod
    def config_file(cls, topic):
        return osp.join(ROOT_PATH, f"{topic.lower()}.config.yaml")

    def ensure_file(self, topic):
        """Generate the config file if it does not exist."""
        config_file = self.config_file(topic)
        if not osp.isfile(config_file):
            shutil.copyfile(osp.join(
                self._abs_dirpath, f"configs/{osp.basename(config_file)}"),
                config_file)

    def load(self, det, topic, **kwargs):
        """Update configs from the config file.

        :param str det: detector name.
        :param str topic: topic name
        """
        self.ensure_file(topic)

        self.__setitem__("DETECTOR", det)
        self.__setitem__("TOPIC", topic)

        # update the configurations of the main detector
        for k, v in self._detector_config[det]._asdict().items():
            self[k] = v

        self.from_file(det, topic)

        # kwargs may contain the configuration for the main detector
        for k, v in kwargs.items():
            if k in self:
                self.__setitem__(k, v)
            else:
                raise KeyError

    def from_file(self, detector, topic):
        """Read configs from the config file."""
        config_file = self.config_file(topic)
        with open(config_file, 'r') as fp:
            try:
                cfg = yaml.load(fp, Loader=yaml.Loader)
            except (ScannerError, ParserError) as e:
                msg = f"Invalid config file: {config_file}\n{repr(e)}"
                logger.error(msg)
                raise OSError(msg)

        if cfg is None:
            raise ValueError(f"Config file {config_file} is empty: "
                             f"returned 'None' from the loader!")

        # update the main detector config
        # TODO: may add "MODULE_POSITIONS" in the future
        det_cfg = cfg.get("DETECTOR", dict()).get(detector, dict())
        for key in ["GEOMETRY_FILE", "QUAD_POSITIONS", "BRIDGE_ADDR",
                    "BRIDGE_PORT", "LOCAL_ADDR", "LOCAL_PORT",
                    "SAMPLE_DISTANCE", "PHOTON_ENERGY"]:
            if key in det_cfg:
                # convert 'GEOMETRY_FILE' to absolute path if it is not
                if key == "GEOMETRY_FILE":
                    geom_path = det_cfg[key]
                    if not osp.isabs(geom_path):
                        det_cfg[key] = osp.join(
                            self._abs_dirpath, 'geometries', geom_path)
                elif key == "QUAD_POSITIONS":
                    pos = det_cfg[key]
                    det_cfg[key] = ((pos['x1'], pos['y1']),
                                    (pos['x2'], pos['y2']),
                                    (pos['x3'], pos['y3']),
                                    (pos['x4'], pos['y4']))

                self[key] = det_cfg[key]

        # update data sources
        src_cfg = cfg.get("SOURCE", dict())

        self["SOURCE_DEFAULT_TYPE"] = src_cfg["DEFAULT_TYPE"]
        for ctg, srcs in src_cfg.get("CATEGORY", dict()).items():
            if ctg == "JungFrauPR":
                logger.warning(
                    f"Found 'JungFrauPR' in {config_file}\n\n"
                    f"Please modify your config file: \n"
                    f"1. Move non-duplicated data sources under 'JungFrauPR' "
                    f"into 'JungFrau';\n"
                    f"2. Remove JungFrauPR from 'SOURCE' and 'DETECTOR'.\n")

                if "JungFrau" not in src_cfg.get("CATEGORY", dict()):
                    # try to import sources under JungFrauPR later
                    ctg = "JungFrau"

            # We assume the main detector type is exclusive, i.e., only one
            # of them will be needed.
            if ctg == detector or ctg in self._misc_source_categories:
                self.control_sources[ctg] = srcs.get("CONTROL", dict())
                self.pipeline_sources[ctg] = srcs.get("PIPELINE", dict())

                if ctg == "JungFrau":
                    # try to merge "JungFrauPR"
                    self.pipeline_sources[ctg].update(
                        src_cfg.get("CATEGORY", dict()).get(
                            "JungFrauPR", dict()).get(
                            "PIPELINE", dict()))

            elif ctg not in self.detectors() and ctg not in ("JungFrauPR",):
                raise ValueError(
                    f"Invalid source category: {ctg}!\n"
                    f"The valid source categories are: "
                    f"{self._misc_source_categories + self.detectors()}")

        # update connection
        con_cfg = cfg.get("STREAMER", dict())
        stream_cfg = con_cfg.get("ZMQ", dict())
        for name, opts in stream_cfg.items():
            self.appendix_streamers.append(
                self.StreamerEndpointItem(
                    name, opts["DEFAULT_TYPE"], opts["ADDR"], opts["PORT"]))


class ConfigWrapper(abc.Mapping):
    """Readonly config."""
    def __init__(self):
        super().__init__()
        self._data = _Config()

    def __contains__(self, key):
        """Override."""
        return self._data.__contains__(key)

    def __getitem__(self, key):
        """Override."""
        return self._data.__getitem__(key)

    def __len__(self):
        """Override."""
        return self._data.__len__()

    def __iter__(self):
        """Override."""
        return self._data.__iter__()

    def load(self, detector, topic, **kwargs):
        self._data.load(detector, topic, **kwargs)

    @property
    def detectors(self):
        return _Config.detectors()

    @property
    def topics(self):
        return _Config.topics()

    @property
    def config_file(self):
        topic = self._data['TOPIC']
        if topic:
            return _Config.config_file(topic)
        raise ValueError("TOPIC is not specified!")

    @property
    def setup_file(self):
        detector = self._data['DETECTOR']
        return osp.join(ROOT_PATH, f".{detector.lower()}.setup.yaml")

    @property
    def control_sources(self):
        return self._data.control_sources

    @property
    def pipeline_sources(self):
        return self._data.pipeline_sources

    @property
    def appendix_streamers(self):
        return self._data.appendix_streamers

    @staticmethod
    def parse_detector_name(detector):
        if detector == 'JUNGFRAU':
            return 'JungFrau'

        if detector == 'FASTCCD':
            return 'FastCCD'

        if detector == 'BASLERCAMERA':
            return 'BaslerCamera'

        if detector == 'EPIX100':
            return 'ePix100'

        return detector.upper()


config = ConfigWrapper()  # global configuration
